prompting/validators/criteria.py

```diff
@@ -0,0 +1,116 @@
+    def evaluate(self, completions: List[str]) -> torch.FloatTensor:
+        pass
+
+    @abstractmethod
+    def compose_text(self) -> str:
+        pass
+
+
+class TextLengthUnitEnum(Enum):
+    CHARACTERS = "characters"
+    WORDS = "words"
+    SENTENCES = "sentences"
+    PARAGRAPHS = "paragraphs"
+
+
+@dataclass
+class MatchLengthCriteria(TaskCriterion):
+    text: str = "Your response must have {target_length} {unit}."
+    penalty: float = 0.1
+    target_length: int = 100
+    unit: TextLengthUnitEnum = TextLengthUnitEnum.WORDS
+
+    def _count_sentences(self, text):
+        # Regex explanation:
+        # \b[A-Z]: Match an uppercase letter at the beginning of a word (assumed start of a sentence).
+        # (?:[a-zA-Z]\. ){0,2}: Match 0-2 occurrences of an abbreviation-like pattern (single letter followed by a dot and space).
+        # (?:[a-zA-Z]+\s+){1,}: Match at least one word followed by one or more whitespace characters.
+        # [a-zA-Z]+: Match the last word before the end punctuation.
+        # [.?!]: Match sentence-ending punctuation.
+        # (?!\w): Negative lookahead to ensure the punctuation is not followed by an alphanumeric character (part of an abbreviation).
+        # The pattern ignores common abbreviations by not counting them as sentence terminators.
+        pattern = r"\b[A-Z](?:[a-zA-Z]\. ){0,2}(?:[a-zA-Z]+\s+){1,}[a-zA-Z]+[.?!](?!\w)"
+
+        # Find all matches
+        sentences = re.findall(pattern, text)
+
+        # Return the count
+        return len(sentences)
+
+    def _get_completion_length(self, response: str) -> int:
+        unit_to_split_pattern = {
+            TextLengthUnitEnum.CHARACTERS: None,
+            TextLengthUnitEnum.SENTENCES: None,
+            TextLengthUnitEnum.WORDS: r"\s+",
+            TextLengthUnitEnum.PARAGRAPHS: r"\n\n+",
+        }
+
+        if self.unit == TextLengthUnitEnum.CHARACTERS:
+            return len(response)
+        elif self.unit == TextLengthUnitEnum.SENTENCES:
+            return self._count_sentences(response)
+        else:
+            split_pattern = unit_to_split_pattern[self.unit]
+            return len(re.split(split_pattern, response.strip()))
+
+    def evaluate(self, completions: List[str]) -> torch.FloatTensor:
+        penalties = torch.zeros(len(completions), dtype=torch.float32)
+        for idx, completion in enumerate(completions):
+            completion_length = self._get_completion_length(completion)
+            if completion_length != self.target_length:
+                # Scales the penalty using a quadratic function based on the deviation of the response length from the target length.
+                # The penalty is designed to be gentler on smaller deviations and steeper on larger deviations from the target length.
+                # The computed penalty is capped between 0 and 1.
+                penalty_scale_factor = np.abs(
+                    1 - (completion_length / self.target_length) ** 2
```

Comment:
change to gaussian

Reply:
oh ok, so instead of 

```python
penalty_scale_factor = np.abs(1 - (completion_length / self.target_length) ** 2).clip(0, 1)
```

you want 

```python
penalty_scale_factor = 1 - np.exp(-10 * (completion_length / self.target_length) ** 2)
```

is that correct?

Reply:
not quite. 
```python
relative_error = (self.target_length - completion_length) / self.target_length
penalty_scale_factor = 1 - np.exp(-10 * relative_error**2)
```